File: algorithm/src/main/java/com/example/algorithm/context/DataContext.java
package com.example.algorithm.context;

import com.example.algorithm.entity.AlternativeEntity;
import com.example.algorithm.entity.AlternativePair;
import com.example.algorithm.entity.CriteriaEntity;
import com.example.algorithm.entity.Rule;
import com.example.algorithm.entity.RuleSet;
import lombok.Data;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Data
public class DataContext {

    private List<AlternativeEntity> alts;
    private List<CriteriaEntity> criterias = new ArrayList<>();
    private List<Rule> P = new ArrayList<>();
    private List<Rule> I = new ArrayList<>();

    public CriteriaEntity getCriteriaByNameOrNull(String name) {
        for (var criteria : criterias) {
            if (criteria.getName().equals(name)) {
                return criteria;
            }
        }
        return null;
    }

    public Set<String> getCriteriaNames() {
        var result = new HashSet<String>();
        for (var criteria : criterias) {
            result.add(criteria.getName());
        }
        return result;
    }

    public void addP(AlternativePair statement) {
        P.add(new Rule(statement, RuleSet.PREPARE));
    }

    public void addI(AlternativePair statement) {
        I.add(new Rule(statement, RuleSet.EQUAL));
    }

    public void removeRule(Rule rule) {
        if (rule.getSet() == RuleSet.EQUAL) {
            I.remove(rule);
        } else {
            P.remove(rule);
        }
    }

    public void addCriteria(CriteriaEntity criteria) {
        criterias.add(criteria);
    }
}

File: shared/src/commonMain/kotlin/DataContext.kt
package org.example

import kotlinx.serialization.Serializable

@Serializable
data class DataContext(
    var alts: MutableList<AlternativeEntity>,
    var criterias: MutableList<CriteriaEntity>,
    var P: MutableList<RuleEntity>,
    var I: MutableList<RuleEntity>,
    var altsComparsion: MutableList<AlternativeComparsionEntity>
) {
    fun getCriteriaByNameOrNull(name: String): CriteriaEntity? {
        return criterias.firstOrNull { criteria ->
            criteria.name == name
        }
    }

    fun getCriteriaNames(): Set<String> {
        return criterias.map { it.name }.toSet()
    }

    fun getNonPriorAlts(): List<AlternativeEntity> {
        val notPrepareAlts = altsComparsion
            .filter { it.rule.set == RuleSet.PREPARE }
            .map { it.rule.pair.second }
        return alts.filter { alt -> notPrepareAlts.firstOrNull { it.isEqual(alt) } == null }
    }

    fun addP(statement: AlternativePair) {
        P.add(RuleEntity(statement, RuleSet.PREPARE))
    }

    fun addI(statement: AlternativePair) {
        I.add(RuleEntity(statement, RuleSet.EQUAL))
    }

    fun removeRule(rule: RuleEntity) {
        when (rule.set) {
            RuleSet.PREPARE -> P.remove(rule)
            RuleSet.EQUAL -> I.remove(rule)
        }
    }

    fun addCriteria(criteria: CriteriaEntity) {
        criterias.add(criteria)
    }

    fun removeCriteria(criteria: CriteriaEntity) {
        criterias.remove(criteria)
    }

    fun formRulesByCriteriaValues() {
        P.removeAll { it.pair.first.criteriaToValue.count() == 1 }
        I.removeAll { it.pair.first.criteriaToValue.count() == 1 }
        criterias.forEach { criteria ->
            val priorToValue = criteria.valueToPrior
                .map { it.value to it.key }
                .sortedBy { it.first }
            priorToValue.forEachIndexed { index, pair ->
                for (i in index + 1 until priorToValue.size) {
                    val alternative1 = AlternativeEntity(-1, String(), mapOf(criteria.name to pair.second))
                    val alternative2 = AlternativeEntity(
                        -1, String(), mapOf(criteria.name to priorToValue.get(i).second))
                    if (priorToValue.get(i).first == pair.first) {
                        addI(AlternativePair(alternative1, alternative2))
                    } else {
                        addP(AlternativePair(alternative1, alternative2))
                    }
                }
            }
        }
    }
}
